ROX-34822: Add operator ARCHITECTURE.md (#20785)
stackrox/stackrox: operator/ARCHITECTURE.md

# Operator Architecture

This document describes the internal architecture of the StackRox operator.
For how to extend the CRDs and set defaults, see [EXTENDING_CRDS.md](EXTENDING_CRDS.md) and [DEFAULTING.md](DEFAULTING.md).

## High-level design

The operator is a hybrid Go+Helm-based operator built on a
[fork](https://github.com/stackrox/helm-operator) of
[operator-framework/helm-operator-plugins](https://github.com/operator-framework/helm-operator-plugins).
The fork includes features and bug fixes specific to StackRox that have not
been upstreamed; see [ROX-7911](https://issues.redhat.com/browse/ROX-7911).

> It watches two CRDs — `Central` and `SecuredCluster` — and for each, renders

Reviewer: nit

```suggestion
It watches essentially two CRDs — `Central` and `SecuredCluster` — and for each, renders
```

For completeness, since it watches also a couple of other resources...

profile from `apiserver.config.openshift.io/cluster`. This profile influences:

- The operator's own metrics server TLS configuration.
- Operand TLS settings (injected via the TLS profile enricher).
- A [watcher](internal/tlsprofile/watch.go) that triggers a graceful shutdown of the operator process
  (by cancelling the manager's root context) when the cluster TLS profile
  changes, so that it restarts with the updated profile.

### Proxy handling ([`internal/proxy/`](internal/proxy/))

The operator captures proxy environment variables (`HTTP_PROXY`, `HTTPS_PROXY`,
etc.) at startup and:

- Creates a Secret per CR containing proxy env vars (via a [pre-extension](internal/proxy/extension.go)).
- Injects the proxy settings into Helm values (via an [enricher](internal/proxy/translation.go)).

### Cross-CR interaction

Central and SecuredCluster reconcilers watch each other's CR type. This enables:

- **Central** to auto-generate init bundles when a SecuredCluster appears in
  the same namespace.
- **SecuredCluster** to decide whether to deploy a local scanner based on
  whether a Central exists in the same namespace.

## Directory layout

> ```

Author: ```suggestion
```text
```

operator/
├── api/v1alpha1/          CRD Go types, defaults merging, deep-copy
├── cmd/                   Entry point (main.go)
├── internal/
│   ├── central/           Central-specific logic
│   │   ├── carotation/      CA certificate rotation
│   │   ├── defaults/        Static and dynamic defaulting flows
│   │   ├── extensions/      Pre-extensions (secrets, PVCs, TLS, defaulting, collision check)
│   │   ├── reconciler/      Wires up the Central Helm reconciler
│   │   └── values/          CR-to-Helm-values translation
│   ├── securedcluster/    SecuredCluster-specific logic (mirrors central/ structure)
│   │   ├── defaults/        Static and dynamic defaulting flows
│   │   ├── extensions/      Pre-extensions (cluster name, passwords, defaulting)
│   │   ├── reconciler/      Wires up the SecuredCluster Helm reconciler
│   │   ├── scanner/         Local scanner auto-sense logic
│   │   └── values/          CR-to-Helm-values translation
│   ├── common/            Shared logic

> │   │   ├── confighash/      Config-hash computation and pod-template annotation

Reviewer 2: ```suggestion
│   │   ├── confighash/      CA certificate hash and pod-template annotation for rollout on CA rotation
```

nit: this could be a bit more concrete

Reviewer 2: I think this got reverted when you forced-pushed. 😅

Author: Indeed, thanks for spotting this! https://github.com/stackrox/stackrox/pull/21054 is out to bring this and more in
